text/0008-kadalu-kubernetes-storage-monitoring.adoc

```diff
@@ -0,0 +1,153 @@
+
+== Authors
+
+- Aravinda Vishwanathapura <aravinda@kadalu.io>
+
+
+== Motivation
+
+- Business logic in one place - If the metrics are exported by individual pods then business logic should be added in the Prometheus server to make decisions.
+- Get all metrics in one place.
+- Events and Alerts - Kadalu Operator can generate events and alerts based on the metrics collected.
+- Kadalu Operator gets more power - Kadalu Operator can take better decisions for the internal maintenance activities depending on the Cluster state. For example, smart rolling upgrade based on Heal pending and Server pods health metrics.
+- Supports non-Prometheus consumers.
+- Additional information export for non-Prometheus API users - For example, list of Volume options configured, Version information etc.
+- External monitoring integration is easy - Since all metrics are available in single place, it will be easy to integrate the metrics with external Cloud hosted monitoring solutions.
+
+== Detailed design
+
+Kadalu Operator runs `kubectl get pods -n kadalu` to get the list of all the resources available in Kadalu namespace. Additionally it fetches the nodes list and all the Storage information from the ConfigMap. With these information a few metrics will be derived as follows.
+
+- Number of Up CSI node plugins by comparing the list of nodes and the list returned by `get pods` command.
+- Pod health based on "Running" state - Up if all the containers in the Pod are up, Partial if only a few containers in the Pod are Up. Down if all Containers are not Running. Operator also knows the expected containers or processes, based on the state of the Pod it also marks the state for respective processes. For example, if a server pod is down, then mark server, self heal daemon and exporter states as down. Type of Storage will be fetched from the ConfigMap, self heal daemon is marked as down only if the Volume type is Replicate or Disperse Volume.
+- Pod Uptime in seconds.
+
+Instead of exporting the metric with Pod name, Kadalu Operator organizes the metrics as proper hierarchy. For example, Metrics from the server pods will be added to the respective brick in the Volume info.
+
+From the collected list of Pod IPs, call HTTP API(`GET /_api/metrics`) for each IPs and collect the metrics(Only if Pod state is Running). If any IP produces Connection refused errors then mark that Pod's health as Down/Unknown.
+
+Once the metrics are collected from all the Pods, post process the metrics and re-expose them under single API.
+
+.Prometheus Metrics
+----
+GET /metrics
+----
+
+.JSON output
+----
+GET /metrics.json
+----
+
+=== Types of Metrics
+
+==== Count Metrics
+
+Metrics related to the resource counts.
+
+- Number of Storages
+- Number of PVs
+- Number of Pods in the Kadalu namespace.
+
+==== Health Metrics
+
+Metrics related to the state of the resources.
```

Comment:
- Does this mean we'll make data available to the user from which below states can be inferred?
- Same for remaining `How to` questions